src/transaction_builder.cpp

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) 2018 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "transaction_builder.h"
+
+#include "main.h"
+#include "pubkey.h"
+#include "script/sign.h"
+
+#include <boost/variant.hpp>
+#include <librustzcash.h>
+
+SpendDescriptionInfo::SpendDescriptionInfo(
+    libzcash::SaplingExpandedSpendingKey xsk,
+    libzcash::SaplingNote note,
+    uint256 anchor,
+    ZCSaplingIncrementalWitness witness) : xsk(xsk), note(note), anchor(anchor), witness(witness)
+{
+    librustzcash_sapling_generate_r(alpha.begin());
+}
+
+TransactionBuilder::TransactionBuilder(
+    const Consensus::Params& consensusParams,
+    int nHeight,
+    CKeyStore* keystore) : consensusParams(consensusParams), nHeight(nHeight), keystore(keystore)
+{
+    mtx = CreateNewContextualCMutableTransaction(consensusParams, nHeight);
+}
+
+bool TransactionBuilder::AddSaplingSpend(
+    libzcash::SaplingExpandedSpendingKey xsk,
+    libzcash::SaplingNote note,
+    uint256 anchor,
+    ZCSaplingIncrementalWitness witness)
+{
+    // Consistency check: all anchors must equal the first one
+    if (!spends.empty()) {
+        if (spends[0].anchor != anchor) {
+            return false;
+        }
+    }
+
+    spends.emplace_back(xsk, note, anchor, witness);
+    mtx.valueBalance += note.value();
+    return true;
+}
+
+void TransactionBuilder::AddSaplingOutput(
+    libzcash::SaplingFullViewingKey from,
+    libzcash::SaplingPaymentAddress to,
+    CAmount value,
+    std::array<unsigned char, ZC_MEMO_SIZE> memo)
+{
+    auto note = libzcash::SaplingNote(to, value);
+    outputs.emplace_back(from.ovk, note, memo);
+    mtx.valueBalance -= value;
+}
+
+void TransactionBuilder::AddTransparentInput(COutPoint utxo, CScript scriptPubKey, CAmount value)
+{
+    if (keystore == nullptr) {
+        throw std::runtime_error("Cannot add transparent inputs to a TransactionBuilder without a keystore");
+    }
+
+    mtx.vin.emplace_back(utxo);
+    tIns.emplace_back(scriptPubKey, value);
+}
+
+bool TransactionBuilder::AddTransparentOutput(CTxDestination& to, CAmount value)
+{
+    if (!IsValidDestination(to)) {
+        return false;
+    }
+
+    CScript scriptPubKey = GetScriptForDestination(to);
+    CTxOut out(value, scriptPubKey);
+    mtx.vout.push_back(out);
+    return true;
+}
```

Comment:
Should there be a similar method that just takes a ``CTxOut``, in case a caller wants to use a funky script?

Reply:
If someone needs that in future, they can add it.